gobblin-runtime/src/main/java/org/apache/gobblin/runtime/messaging/hdfs/FileSystemMessageBuffer.java

```diff
@@ -86,39 +103,77 @@ private void scheduleTaskForPollingNewMessages() {
   }
 
   private List<DynamicWorkUnitMessage> getNewMessages() {
-    // STUB: TODO GOBBLIN-1685 Read all files in dir, deserialize, then clean up byte files in dir
-    return null;
+    List<DynamicWorkUnitMessage> messageList = new LinkedList<>();
+    try {
+      FileStatus[] fileStatus = fs.listStatus(path);
+      for(FileStatus status : fileStatus){
+        Path singlePath = status.getPath();
+        if (fs.isFile(singlePath))
+        {
+          FSDataInputStream fis = fs.open(singlePath);
+          DynamicWorkUnitMessage message = DynamicWorkUnitSerde.deserialize(ByteStreams.toByteArray(fis));
+          messageList.add(message);
+          fs.delete(singlePath, true);
+          log.debug("Fetched message from HDFS path {}.", singlePath.toString());
+        }
+      }
+    } catch (IOException ioException) {
+      log.error("Failed to get message from HDFS path {}.", path.toString(), ioException);
+    }
+
+    return messageList;
+  }
+
+  private void persistMessage(byte[] serializedMsg, DynamicWorkUnitMessage message) {
+    try (FSDataOutputStream os = fs.create(this.getFilePathForMessage(message), true)) {
+      os.write(serializedMsg);
+      log.info("Persisted message into HDFS path {} for workunit {}.", path.toString(), message.getWorkUnitId());
+    } catch (IOException ioException) {
+      log.error("Failed to persist message into HDFS path {}.", path.toString(), ioException);
+    }
   }
 
-  private void persistMessage(byte[] serializeMsg) throws IOException {
-    // STUB: TODO GOBBLIN-1685 write serialized message to FileSystem
+  private Path getFilePathForMessage(DynamicWorkUnitMessage message) {
+    String messageFileName = message.getWorkUnitId() + "-" + System.currentTimeMillis();
+    Path messagePath = PathUtils.mergePaths(path, new Path(messageFileName));
+    return PathUtils.addExtension(messagePath, messageFileSuffix);
```

Comment:
Is the file actually a json? We serialize first to bytes, so I am not sure that the file is still a json. Maybe it's better to give it a different file extension.

Reply:
changed to .wumessage, as corresponding to .wu for workunit